core/src/main/java/com/scalar/db/storage/cosmos/CosmosTableMetadata.java

```diff
@@ -2,52 +2,58 @@
 
 import com.google.common.base.MoreObjects;
 import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import java.util.Collections;
 import java.util.LinkedHashSet;
-import java.util.List;
 import java.util.Map;
 import java.util.Objects;
 import java.util.Set;
-import javax.annotation.concurrent.NotThreadSafe;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
 
 /**
  * A metadata class for a table of ScalarDB to know the type of each column
  *
  * @author Yuji Ito
  */
 @SuppressFBWarnings({"EI_EXPOSE_REP", "EI_EXPOSE_REP2"})
-@NotThreadSafe
+@Immutable
 public class CosmosTableMetadata {
-  private String id;
-  private LinkedHashSet<String> partitionKeyNames;
-  private LinkedHashSet<String> clusteringKeyNames;
-  private Map<String, String> clusteringOrders;
-  private Set<String> secondaryIndexNames;
-  private Map<String, String> columns;
-
-  public CosmosTableMetadata() {}
-
-  public void setId(String id) {
-    this.id = id;
-  }
-
-  public void setPartitionKeyNames(List<String> partitionKeyNames) {
-    this.partitionKeyNames = new LinkedHashSet<>(partitionKeyNames);
-  }
-
-  public void setClusteringKeyNames(List<String> clusteringKeyNames) {
-    this.clusteringKeyNames = new LinkedHashSet<>(clusteringKeyNames);
-  }
-
-  public void setClusteringOrders(Map<String, String> clusteringOrders) {
-    this.clusteringOrders = clusteringOrders;
+  private final String id;
+  private final LinkedHashSet<String> partitionKeyNames;
+  private final LinkedHashSet<String> clusteringKeyNames;
+  private final Map<String, String> clusteringOrders;
+  private final Set<String> secondaryIndexNames;
+  private final Map<String, String> columns;
+
+  public CosmosTableMetadata() {
+    this(null, null, null, null, null, null);
   }
 
-  public void setSecondaryIndexNames(Set<String> secondaryIndexNames) {
-    this.secondaryIndexNames = secondaryIndexNames;
+  public CosmosTableMetadata(
+      @Nullable String id,
+      @Nullable LinkedHashSet<String> partitionKeyNames,
```

Comment:
This can be declared as an interface `Set<String>`.

Reply:
Ah no, we can't do that because the insertion order is important for this.

Reply:
I see. Understood. (Another option might be simply using `List` with de-duplications)